ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -575,19 +621,68 @@ KEXT_STATIC int HandleVnodeOperation(
+                /* At this stage we know the file is NOT empty, but it could be
+                 * in either placeholder (hydrated) or full (modified/added to
+                 * watchlist) state.
+                 *
+                 * If it's a placeholder, we need to notify the provider before
+                 * allowing modifications so that it is tracked as a full file
+                 * (and the placeholder xattr is removed).
+                 * If the provider is offline in this situation, we want to
+                 * prevent writes (by most processes) to the file so that the
+                 * modifications don't go undetected.
+                 *
+                 * If it's already a full file on the other hand, nothing needs
+                 * to be done, so we can allow writes even when the provider is
+                 * offline.
+                 */
+                ProviderStatus providerStatus;
+                bool shouldMessageProvider = TryGetVirtualizationRoot(
+                    &perfTracer,
+                    context,
+                    currentVnode,
+                    pid,
+                    CallbackPolicy_UserInitiatedOnly,
+                    // At this stage, we don't yet know if the file is still a placeholder, so don't deny yet even if offline
+                    false, // denyIfOffline
+                    &root,
+                    &vnodeFsidInode,
+                    &kauthResult,
+                    kauthError,
+                    &providerStatus);
+                
+                if (!shouldMessageProvider)
                 {
-                    goto CleanupAndReturn;
+                    // If we don't need to message the provider, that's normally
+                    // the end of it, but in the case where the provider is
+                    // offline we need to do further checks.
+                    if (providerStatus != Provider_IsOffline)
+                    {
+                        goto CleanupAndReturn;
+                    }
                 }
                 
                 PrjFSFileXAttrData rootXattr = {};
                 SizeOrError xattrResult = Vnode_ReadXattr(currentVnode, PrjFSFileXAttrName, &rootXattr, sizeof(rootXattr));
                 if (xattrResult.error == ENOATTR)
                 {
-                    // Only notify if the file is still a placeholder
+                    // If the file does not have the attribute, this means it's
+                    // already in the "full" state, so no need to send a provider
+                    // message or block the I/O if the provider is offline
                     goto CleanupAndReturn;
                 }
                 
+                if (providerStatus == Provider_IsOffline)
+                {
+                    assert(!shouldMessageProvider);
+                    // Deny write access to placeholders in an offline root
+                    kauthResult = KAUTH_RESULT_DENY;
```

Comment:
I *think* this is OK but it would be worth a second pair of eyes.  
At this point we are ignoring any checks from TryGetVirtualizationRoot, and making the decision to deny solely on if we have a placeholder and we're offline.  That makes snese to me.

Reply:
Note that `TryGetVirtualizationRoot` will *never* return `true` in the case of an offline provider, so there's not really anything to ignore.

Reply:
@pmj what I meant is TryGetVirtualizationRoot could return `false` for a reason other than offline provider.

Reply:
Yes, and that's already caught earlier on, on lines 656-662. (Of course if by "ignoring any checks" you mean we've already dealt with them then we're just vigorously agreeing with one another.)

Reply:
I think we are :-)